uber/site_sections/budget.py

```diff
@@ -139,44 +142,89 @@ def generate_promo_codes(
         except:
             is_single_promo_code = 0
 
+        result = {
+            'message': message,
+            'promo_codes': [],
+            'words': [(i, s) for (i, s) in words.items()],
+            'is_single_promo_code': is_single_promo_code,
+            'code': code,
+            'count': count,
+            'use_words': use_words,
+            'length': length,
+            'segment_length': segment_length,
+            'expiration_date': expiration_date,
+            'discount_type': discount_type,
+            'discount': discount,
+            'uses_allowed': uses_allowed
+        }
+
         if cherrypy.request.method == 'POST':
             codes = None
             if is_single_promo_code:
-                if params.get('code', '').strip():
-                    codes = [params['code']]
-                else:
-                    count = 1
+                count = 1
+                if code:
+                    codes = [code]
 
             if use_words and not codes and \
                     not any(s for (_, s) in words.items()):
-                raise HTTPRedirect('generate_promo_codes?message='
-                    'Please add some promo code words!')
+                result['message'] = 'Please add some promo code words!'
+                return result
 
             if not codes:
                 if use_words:
                     codes = PromoCode.generate_word_code(count)
                 else:
-                    length = int(params.get('length', 12))
-                    segment_length = int(params.get('segment_length', 3))
+                    try:
```

Comment:
I can see you were trying to make sure length and segment_length were ints, but I think putting in a non-integer is a check-failure-worthy admin error. Instead of doing this change, I'd recommend adding model validations to return "What you entered for {} isn't even a number". That way if someone for some reason enters the word 'four' we don't generate 50 promo codes of length three unexpectedly.

I can see that you don't run the check function until later, and you want to generate the random codes now. I'm not sure exactly how to handle that... it seems like those model checks are being done way too late.